moment_kinetics/src/numerical_dissipation.jl

```diff
@@ -41,65 +42,54 @@ vz_dissipation_coefficient
-                                     neutral_input::Dict, is_1V)
-    if is_1V && "vpa_dissipation_coefficient" ∈ keys(ion_input)
-        # Set default for vz_dissipation_coefficient the same as
-        # ion_vpa_dissipation_coefficient for 1V case
-        neutral_input["vz_dissipation_coefficient"] =
-            get(neutral_input, "vz_dissipation_coefficient",
-                ion_input["vpa_dissipation_coefficient"])
+function setup_numerical_dissipation(input_dict)
+    ion_settings = set_defaults_and_check_section!(
+        input_dict, "ion_numerical_dissipation";
+        vpa_boundary_buffer_damping_rate=-1.0,
+        vpa_boundary_buffer_diffusion_coefficient=-1.0,
+        vpa_dissipation_coefficient=-1.0,
+        vperp_dissipation_coefficient=-1.0,
+        z_dissipation_coefficient=-1.0,
+        r_dissipation_coefficient=-1.0,
+        moment_dissipation_coefficient=-1.0,
+        force_minimum_pdf_value=-Inf,
+       )
+    ion_settings = deepcopy(ion_settings)
+    if ion_settings["force_minimum_pdf_value"] == -Inf
+        ion_settings["force_minimum_pdf_value"] = nothing
     end
+    ion_params = Dict_to_NamedTuple(ion_settings)
+    electron_settings = set_defaults_and_check_section!(
+        input_dict, "electron_numerical_dissipation";
+        vpa_boundary_buffer_damping_rate=-1.0,
+        vpa_boundary_buffer_diffusion_coefficient=-1.0,
+        vpa_dissipation_coefficient=-1.0,
+        vperp_dissipation_coefficient=-1.0,
+        z_dissipation_coefficient=-1.0,
+        r_dissipation_coefficient=-1.0,
+        moment_dissipation_coefficient=-1.0,
+        force_minimum_pdf_value=-Inf,
+       )
+    electron_settings = deepcopy(electron_settings)
+    if electron_settings["force_minimum_pdf_value"] == -Inf
+        electron_settings["force_minimum_pdf_value"] = nothing
+    end
+    electron_params = Dict_to_NamedTuple(electron_settings)
+    neutral_settings = set_defaults_and_check_section!(
+        input_dict, "neutral_numerical_dissipation";
+        vz_dissipation_coefficient=-1.0,
+        z_dissipation_coefficient=-1.0,
+        r_dissipation_coefficient=-1.0,
+        moment_dissipation_coefficient=-1.0,
+        force_minimum_pdf_value=-Inf,
+       )
+    neutral_settings = deepcopy(neutral_settings)
+    if neutral_settings["force_minimum_pdf_value"] == -Inf
+        neutral_settings["force_minimum_pdf_value"] = nothing
+    end
+    neutral_params = Dict_to_NamedTuple(neutral_settings)
 
-    ion_input_dict = Dict(Symbol(k)=>v for (k,v) in ion_input)
-    ion_params = ion_num_diss_params(; ion_input_dict...)
-    electron_input_dict = Dict(Symbol(k)=>v for (k,v) in electron_input)
-    electron_params = electron_num_diss_params(; electron_input_dict...)
-    neutral_input_dict = Dict(Symbol(k)=>v for (k,v) in neutral_input)
-    neutral_params = neutral_num_diss_params(; neutral_input_dict...)
-
-    return numerical_dissipation_parameters(ion_params, electron_params, neutral_params)
```

Comment:
`NamedTuple` is immutable.

I want to use `set_defaults_and_check_section!()` to read the inputs, because that's the function that registers the section as being 'expected' so we can check that only expected sections exist. Also, `set_defaults_and_check_section!()` has nicer error messages.

Reply:
I have nothing against `set_defaults_and_check_section!()`. My concern here was just about the object returned by the function. For geometry and composition and etc, we use explicitly defined structs. I find these easier to understand because the typing documentation is explicitly given somewhere. Happy to hear that NamedTuple is immutable, would be even happier if there was an explicit type definition of everything that goes into the NamedTuple (i.e., it was a struct : ) ). This is probably just a preference based on the other existing namelist options.

Reply:
Yeah, there's not quite an ideal solution. I'd like to only 'declare' the options in one place - either as the fields of a struct or as the arguments to `set_defaults_and_check_section!()`. The trouble is that: if you declare the fields and the defaults in a struct definition, then the defaults have to be constants; if you use `get_defaults_and_check_section!()` to get an `OptionsDict` with the right variables, you'd have to declare them twice to have a struct that contains the same variables.

Using `NamedTuple` is a compromise where you do the second option, but have to live with having an implicitly-defined `NamedTuple` rather than an explicitly-defined struct. Sometimes having the names declared twice so you can have the struct will be the best thing to do.

See the new definition of initial conditions in 3ad41d6 for an example of the 'first option'.

Reply:
I acknowledge the two options, and suggest a third -- my preference is for an option always involving an explicitly defined struct, at the expense of declaring the variables twice to get the right defaults, where necessary.

Reply:
That is an option, and we do use it in places. I find the `NamedTuple` clear enough (as the usage is pretty standard, the fields are defined in a call to `set_defaults_and_check_section!()` in a 'setup' function for some module) that it's worth using (sometimes) for the benefit of being easier to update.

I don't think we're going to agree on this. @mabarnes @LucasMontoya4 do you have an opinion? I think to sum up, the question is:
In the case where we want the default for one or more options to not just be a constant value (i.e. depend on another option, or some other parameters) there are two options
1. Define a struct for settings (using `@kwdef`); call `set_defaults_and_check_section!()` with the keyword arguments being the fields of the struct, creating an `OptionsDict`; do some processing to update the default values as necessary, modifying the `OptionsDict`; construct an instance of the struct using the `OptionsDict`.
2. Call `set_defaults_and_check_section!()` with the keyword arguments being the fields of the struct, creating an `OptionsDict`; do some processing to update the default values as necessary, modifying the `OptionsDict`; use `Dict_to_NamedTuple` to create a `NamedTuple` (which is an immutable, implicitly defined struct) from the `OptionsDict`.
1 has an explicitly declared struct, but requires all the fields to be 'defined' twice (once in the struct definition and once in the call to `set_defaults_and_check_section!()`. 2 defines the fields only once, but creates a `NamedTuple` with those fields.

Should we require 1 or 2, or leave it up to whoever is writing the module to decide which seems more convenient?

PS in this particular instance, we could avoid this issue by having constant defaults if we didn't bother to replace `-Inf` with `nothing` for `force_minimum_pdf_value`. I think using `nothing` is marginally more efficient because the `if` can be evaluated at compile-time, but the cost will be totally negligible anyway because it's one float comparison that's outside any loop over arrays. Maybe that would be the best option in this instance!

Reply:
Actually, maybe this is not such a big issue after all. I got into the habit of using `NamedTuple`s when `set_default_and_check_section!()` couldn't handle structs at all, but this discussion with @mrhardman prompteb me to add the version that can directly create an `@kwdef struct` as long as the defaults are constant. Now (with the commits I just pushed) using that version where possible, there are only a few places where we create settings in a `NamedTuple` and:
* `evolve_moments` settings don't propagate far, only until they get put into the `moments` struct.
* output from `get_coordinate_input()` only gets used inside or fed into `define_coordinate()`.
* input for manufactured solutions is only used within that module, and it's useful to make adding new inputs as simple as possible there.
* the `nonlinear_solvers` functionality is still WIP, so it's useful to be able to change it simply (at least for now).
* The reference parameters aren't used in many places, and we might want to add more derived parameters.